book.md

```diff
@@ -398,15 +398,76 @@ cache, without the processes having rights to read or write to the same memory
 locations.  To understand how these techniques work, one needs to understand
 typical organization and operation of a cache.
 
-### Typical cache architecture
-
-
-\missingcontent{Add a description of how a cache operates - as far as is
-necessary to explain the cache covert channel techniques.}
-
-### Flush + Reload
-
-### Prime + Probe
+### Typical CPU cache architecture
+
+There is a wide variety in
+[CPU cache micro-architecture](https://en.wikipedia.org/wiki/CPU_cache) details,
+but the main characteristics that are important to set up a covert channel tend
+to be similar across most popular implementations.
+
+Caches are small and much faster memories than the main memory that aim to keep
+a copy of the data at the most frequently accessed addresses in main memory. The
+set of addresses that are used most frequently changes quickly over time as a
+program executes. Therefore, the addresses that are present in CPU caches also
+evolve quickly over time. The content of the cache may change with every
+executed read or write instruction.
+
+On every read and write instruction, the cache micro-architecture looks up if
+the data for the requested address happens to be present in the cache. If it is,
+the CPU can continue executing quickly; if not, dependent operations will have
+to wait until the data returns from the much more slowly main memory.
+Typical access times is 3 to 5 CPU cycles for the fastest cache on a CPU versus
+hundreds of cycles for a main memory access.\index{memory access time}
+
+Most systems have multiple level of cache\index{multi-level cache}, each with a
+different trade-off between cache size\index{cache size} and access
+time\index{cache access time}. Some typical characteristics might be:
+
+* A L1 (level 1) cache, 32kB in size, with an access time of 4 cycles.
+* A L2 cache, 256Kb in size, with an access time of 10 cycles.
+* A L3 cache, 16MB in size, with an access time of 40 cycles.
+* Main memory, gigabytes in size, with an access time of 100 cycles.
+
+When the data was not present in the cache, it typically is stored in the cache
+after it's fetched from a slower cache level or main memory. This is often a
+good decision to make as there's a high likelihood the same address will be
+accessed by the program soon after. This high likelihood is known as the
+[principle of locality](https://en.wikipedia.org/wiki/Locality_of_reference)\index{principle
+of locality}\index{locality of reference}.
+
+Data is stored and transferred between cache levels in blocks of aligned memory.
+Such a block is called a \index{cache block} or \index{cache line}. Typical
+sizes are 32, 64 or 128 bytes per cache line.
+
+When data that wasn't previously in the cache needs to be stored in the cache,
+most of the time, room has to be made for it be removing, or
+evicting\index{cache eviction}, some other address/data from it. How that choice
```

Comment:
Nit: could evicting be emphasised, to highlight that you're introducing a common term here? If we decide to change \index, this could be done automatically.

Reply:
Maybe we should write up a writing convention in the appendix that when a term gets introduced it should be typeset in a particular way?

I've factored out this discussion into issue #91.